File: static/CWE-78/python/patched.py
# Remediation for CWE-78: OS Command Injection
# Fix: Pass arguments as a list with shell=False so user input is never interpreted by a shell.

import subprocess
from flask import Flask, request

app = Flask(__name__)


@app.route("/ping")
def ping():
    host = request.args.get("host", "")
    subprocess.run(["ping", "-c", "1", "--", host], shell=False, check=False)
    return "pinged " + host

File: static/CWE-78/javascript/patched.js
// Remediation for CWE-78: Improper Neutralization of Special Elements used in an OS Command ('OS Command Injection')
// Fix: Use execFile with an argument array (no shell) and validate the host so input cannot be interpreted as shell syntax.

const express = require('express');
const { execFile } = require('child_process');

const app = express();

const HOST_PATTERN = /^[a-zA-Z0-9.-]+$/;

app.get('/ping', (req, res) => {
  const host = req.query.host;
  if (typeof host !== 'string' || !HOST_PATTERN.test(host)) {
    return res.status(400).send('invalid host');
  }
  execFile('ping', ['-c', '1', host], (err, stdout, stderr) => {
    if (err) return res.status(500).send(stderr);
    res.send(stdout);
  });
});

app.listen(3000);
